clash-prelude/src/Clash/Annotations/TopEntity.hs

```diff
@@ -206,21 +206,24 @@ See the documentation of 'Synthesize' for the meaning of all its fields.
 
 === 'TestBench' annotation
 
-Tell what binder is the 'TestBench' for a 'Synthesize'-annotated binder.
-
-So in the following example, /f/ has a 'Synthesize' annotation, and /g/ is
-the HDL test bench for /f/.
+Tell what binder is the test bench for a 'Synthesize'-annotated binder.
 
 @
-f :: Bool -> Bool
-f = ...
-{\-\# ANN f (defSyn "f") \#-\}
-{\-\# ANN f (TestBench \'g) \#-\}
+entityBeingTested :: ...
+entityBeingTested = ...
+{\-\# NOINLINE entityBeingTested \#-\}
+{\-\# ANN entityBeingTested (defSyn "entityBeingTested") \#-\}
+
 
-g :: Signal Bool
-g = ...
+myTestBench :: Signal System Bool
+myTestBench = ...
+{\-\# NOINLINE myTestBench \#-\}
+{\-\# ANN myTestBench (TestBench \'entityBeingTested) \#-\}
 @
 
+The 'TestBench' annotation actually already implies a 'Synthesize' annotation on
+the device under test, so the 'defSyn' in the example could have been omitted.
+It will still need the @NOINLINE@ annotation, though.
```

Comment:
I wonder if we should document this sort of semi-magic behaviour.
I think I agree with [Martijn](https://github.com/clash-lang/clash-compiler/issues/1750#issuecomment-814147942) that we might want to remove this (at some point).
And if it's not documented as a feature, it feel like it's easier to remove.

Reply:
I rather think we should warn the user about this rather surprising aspect, and if we replace the functionality by an `isTestBench` annotation, we _really_ need to warn the user explicitly about this change in the Changelog as well as the Haddock documentation. Otherwise people might get very confused why the HDL generation suddenly changes when they add or remove a `TestBench` annotation.

For instance, the following source:
```haskell
module TB where

import Clash.Prelude

f :: Signal System (Unsigned 8) -> Signal System (Unsigned 8)
f = id
{-# NOINLINE f #-}

g :: Signal System Bool
g = (f 5) .==. 5
{-# NOINLINE g #-}
{-# ANN g (TestBench 'f) #-}

h :: Signal System (Unsigned 8) -> Signal System (Unsigned 8)
h = f
{-# NOINLINE h #-}
```

can lead to:
```
$ clash -Wall --vhdl TB -main-is h
[...]
$ ls vhdl/
TB.f  TB.h
```

whereas if we remove that `TestBench` annotation from g (or change it to the proposed `isTestBench`), we suddenly get:
```
clash -Wall --vhdl TB -main-is h
[...]
$ ls -R vhdl/
vhdl/:
TB.h

vhdl/TB.h:
clash-manifest.json  f.vhdl  h.vhdl  TB_h_types.vhdl
```

If we don't document this weird interaction (the user never even asked for `g`, what should its annotations matter?), I think we do our users a disservice.

Reply:
Like @leonschoorl I feel a bit meh about this whole semi-magic behavior, but I think that bolsters the need for documenting it. It's probably best to add a comment that we don't really recommend leaving out `defSyn`. (We already discussed this IRL, but I thought I'd post it here because Leon is out-of-office.)

Reply:
We've been discussing this in the office, and have decided on documenting it _but_ recommending against it explicitly.